### ISSUE TITLE
[Talk] How to write unit test with telegraf?

### ISSUE THREAD

**Issue by b3nab** [telegraf/telegraf#918]
I see #224 but for complex behaviours `telegram-test-api` doesn't work well, especially with a session attached to a telegraf bot.

However, anyone knows how to write tests for a bot without rely on official telegram server?

For example if I want to test the bot offline or I don't want to use my bot that is actually used in production, or I don't want to create two bots (one for prod and one for test), I need something reliable that let me get updates from my bot as if I were using a telegram client. 

There is maybe a method like `bot.getUpdate(message_from_client)` that could help this way?
I mean, just a method that receive the message, do stuffs from middlewares and other things and finally return the message to send to telegram server. 
This way we can test our bots in a detached environment or even offline if we have local databases and this kind of stuff.

**Comment by IMalyugin** [comment]
Wish there was.

At someplaces telegraf doesn’t even handle invalid syntax before sending messages - like inline Queries. If you write them down with a mistake telegram api replies that your query is invalid and quits.

So I believe we are quite a distance away from fully coverable system, and your best bet at the moment would be writing a telegram user based api, that would imitate a user and go with e2e tests.

Personally I got satisfied with just setting my environment straight - using ngrok, child_process, babel-watch and tg-proxy to work with tg bot from localhost.

Basically each time I save any file, half a sec later I can test the result with the bot.

**Comment by aveeday** [comment]
Unit tests should not depend on network. Unit tests should test only one "module" from one "layer" of your application e.g. database, domain, service etc. Module dependencies should be mocked for testing. 

For example:
```
it('should return author name', () => {
  const result = libraryService.query('who wrote Sonnet 18')
  assert(result, 'William Shakespeare')
})
```

Then you can use telegraf api like this:
```
bot.on('text', ctx => ctx.reply(libraryService.query(ctx.message.text)))
```

You should design "modules" for your application to be testable and they will works with any framework.

**Comment by aveeday** [comment]
I found a convenient way to test telegram bots with some kind of integration tests and mocking http response and telegram server api. telegraf implements a bit complicated functionality like scenes and session so it would be nice to have a several integration tests to verify some critical cases. I created an example of how this integration tests can be implemented:
https://github.com/aveeday/telegraf-testing-example

@IMalyugin  could you please provide some examples of invalid syntax? It possible and may be useful to implement a test client that will be validate syntax for use it in integration tests.

**Comment by wojpawlik** [comment]
Duplicate of #224
